### PR TITLE
Paged Attention合入pr，待验证

### PR DESCRIPTION
**Pull request by Susskind115**
Paged Attention合入pr，已修改完当前所有冲突，经验证后方可合入。
特点：vllm-like Scheduler and Memory Manager；Paged Attention；

### CODE REVIEW

File: python/example.py [hunk 1]
```diff
@@ -0,0 +1,158 @@
+import os
+os.environ["CUDA_VISIBLE_DEVICES"] = "0,1,2,3,4,5,6,7"
```
Comment (PanZezhong1725) [inline python/example.py:2]: 这个调脚本的时候用户自己写吧，要么就作为脚本参数传进来

File: scripts/jiuge.py [hunk 2]
```diff
@@ -690,4 +690,4 @@ def test():
 
 
 if __name__ == "__main__":
-    test()
+    test()
```
Comment (PanZezhong1725) [inline scripts/jiuge.py:693]: 这不小心改的？

File: scripts/test_ppl.py [hunk 3]
```diff
@@ -11,17 +11,25 @@
...
     parser.add_argument("--chunk", type=int, default=512)
     args = parser.parse_args()
 
     API_URL = "http://localhost:" + str(args.port) + args.endpoint
     CHUNK_SIZE = args.chunk
-
-    dataset = load_dataset("wikitext", "wikitext-2-raw-v1", split="test")
+    
+    print("Loading dataset...")
+    local_file_paths = {
```
Comment (PanZezhong1725) [inline scripts/test_ppl.py:22]: 这种写死的就不要出现在脚本里了

File: scripts/test_ppl.py [hunk 4]
```diff
@@ -11,17 +11,25 @@
     parser = argparse.ArgumentParser()
     parser.add_argument("--model-path", type=str, required=True)
     parser.add_argument("--port", type=int, default=8000)
-    parser.add_argument("--endpoint", type=str, default="/completions")
```
Comment (PanZezhong1725) [inline scripts/test_ppl.py:14]: ppl用的就是completion接口不是chat

File: setup.sh [hunk 5]
```diff
@@ -0,0 +1,589 @@
+#!/bin/bash
```
Comment (PanZezhong1725) [inline setup.sh:1]: 这是什么？

File: src/tensor.hpp [hunk 6]
```diff
@@ -137,6 +137,12 @@ class Tensor : public std::enable_shared_from_this<Tensor> {
     std::shared_ptr<Tensor> view_as(const std::vector<size_t> &new_shape) const;
     std::shared_ptr<Tensor> view_as(const std::vector<size_t> &new_shape, const std::vector<ptrdiff_t> &new_strides) const;
 
+    // template <typename T>
```
Comment (PanZezhong1725) [inline src/tensor.hpp:140]: 没用就删掉吧

File: src/tensor/tensor.cpp [hunk 7]
```diff
@@ -424,3 +424,61 @@ void Tensor::debug(const std::string &filename) const {
 }
 
 void Tensor::debug() const { this->debug(""); }
+
+
+// template <typename T>
```
Comment (PanZezhong1725) [inline src/tensor/tensor.cpp:429]: 没用就删掉

File: include/infinicore_infer/models/jiuge.h [hunk 8]
```diff
@@ -61,6 +61,24 @@ createJiugeModel(const JiugeMeta *,
...
+// /// @brief 复制 KV Cache
+// __C __export struct KVCache *
+// duplicateKVCache(const struct JiugeModel *,
+//                  const struct KVCache *, uint32_t seq_len);
+
+// /// @brief 销毁 KV Cache
+// __C __export void
+// dropKVCache(const struct JiugeModel *,
+//             struct KVCache *);
+
```
Comment (wooway777) [inline include/infinicore_infer/models/jiuge.h:81]: 是不是不需要加这么多commented code

File: include/infinicore_infer/models/jiuge.h [hunk 9]
```diff
@@ -71,14 +89,19 @@ destroyJiugeModel(struct JiugeModel *);
...
-                struct KVCache **kv_caches,
-                const float *temperature, const uint32_t *topk, const float *topp,
-                uint32_t *output);
+           const uint32_t *tokens, uint32_t ntok,
+           const uint32_t *req_lens, uint32_t nreq, const uint32_t *req_pos,
+           struct KVCache **kv_caches,
+           const int32_t *block_tables,
+           const int32_t *slot_mapping,
+           const float *temperature, const uint32_t *topk, const float *topp,
+           const uint32_t is_prefill, const bool enable_paged_attn,
```
Comment (wooway777) [inline include/infinicore_infer/models/jiuge.h:103]: is_prefill和enable_paged_attn是否都应为bool

File: include/infinicore_infer/models/jiuge.h [hunk 10]
```diff
@@ -87,12 +110,19 @@ inferBatchJiuge(struct JiugeModel *,
...
-                  const uint32_t *tokens, uint32_t ntok,
-                  const uint32_t *req_lens, uint32_t nreq, const uint32_t *req_pos,
-                  struct KVCache **kv_caches,
-                  void *logits);
+             const uint32_t *tokens, uint32_t ntok,
+             const uint32_t *req_lens, uint32_t nreq, const uint32_t *req_pos,
+             struct KVCache **kv_caches,
+             const int32_t *block_tables,
+             const int32_t *slot_mapping,
+             const uint32_t is_prefill, const bool enable_paged_attn,
```
Comment (wooway777) [inline include/infinicore_infer/models/jiuge.h:125]: 同上

File: python/bench.py [hunk 11]
```diff
@@ -0,0 +1,83 @@
+import os
+os.environ["CUDA_VISIBLE_DEVICES"] = "0,1,2,3,4,5,6,7"
+import time
+import sys
+from random import randint, seed
+# from nanovllm import LLM, SamplingParams
```
Comment (wooway777) [inline python/bench.py:6]: 是否需要保留？

File: python/example.py [hunk 12]
```diff
@@ -0,0 +1,158 @@
...
+    # llm = LLM(path, enforce_eager=True, tensor_parallel_size=1, trust_remote_code=True)
+    # path = os.path.expanduser("/home/wanghaojie/vllm/huggingface/9G7B_MHA/")
+    path = args.model_path
+    tokenizer = AutoTokenizer.from_pretrained(path, trust_remote_code=True)
+    llm = LLM(path, device=device_type, enforce_eager=True, 
+              tensor_parallel_size=args.ndev, trust_remote_code=True, 
+              attention_bias=True, enable_paged_attn=args.enable_paged_attn, max_kvcache_tokens=max_kvcache_tokens)
+
+    sampling_params = SamplingParams(temperature=0.6, max_tokens=128)
+    # prompts = [
```
Comment (wooway777) [inline python/example.py:65]: 后面这60行是不是要整理一下
